service/src/test/java/bio/terra/cbas/controllers/TestMethodsApiController.java

```diff
@@ -334,6 +336,32 @@ void validPostRequest() throws Exception {
     assertTrue(newRunSetCaptor.getValue().isTemplate());
   }
 
+  @Test
+  void requestValidationForDuplicateMethod() {
+    initMocks();
+
+    PostMethodRequest duplicateMethodRequest = new PostMethodRequest();
+
+    duplicateMethodRequest.methodName("method1");
+    duplicateMethodRequest.methodDescription("method one");
+    duplicateMethodRequest.methodUrl(validWorkflow);
+    duplicateMethodRequest.methodVersion("v1");
+    duplicateMethodRequest.methodSource(PostMethodRequest.MethodSourceEnum.GITHUB);
+
+    List<String> expectedErrors =
+        new ArrayList<>(List.of("Method method1 already exists. Please select a new method."));
+
+    when(methodDao.getNumberOfMethods(
+            duplicateMethodRequest.getMethodName(),
+            duplicateMethodRequest.getMethodUrl(),
+            duplicateMethodRequest.getMethodVersion(),
+            duplicateMethodRequest.getMethodSource().toString()))
+        .thenReturn(1);
+
+    List<String> actualErrors = MethodsApiController.validateMethod(duplicateMethodRequest);
```

Comment:
There's potentially a super-subtle [TOCTTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) race condition if we use `validateMethod` here. 

Imagine someone writes a buggy script to import workflows and it sends two requests at exactly the same time (or perhaps a user clicks an import button in the UI twice in quick succession). This check for "does it exist" might pass in both threads, and then two instance of the method might still be written to the database in parallel.

The "validate" method was until now mostly concerned with general validity/invalidity of methods, but is now checking validity in a way that changes over time (a workflow which is valid at time t0 might no longer be valid at time t10 or t100).

This isn't particularly bad here (the worst case scenario is both very unlikely to happen, AND is hardly the end of the world!), but it's good to be in the habit of thinking about these things. But something to think about: if the intention is to never allow duplicate entries containing the same set of these values, an alternative would be to create a unique key constraint across those fields (see [example](https://dba.stackexchange.com/questions/316319/how-to-check-record-if-exists-before-insert-with-postgresql)). That way, the insert would fail automatically if another row already contained the values, and no special check would be needed ahead of time.

Reply:
I think this makes complete sense. Thank you for pointing this out. I think it'll be best to create a new ticket for a changelog that adds the `UNIQUE` key constraint on these columns.

Reply:
I am wondering if we can add the `UNIQUE` constraints as part of this PR. The reason is that if we add the constraints in this ticket itself, we might not need to do this validation check in the `validateMethod` as the database itself would throw and error when creating duplicate methods and we can catch that and return a "duplicate method..." error back to the user? Plus it fits within the scope of this ticket itself. What do you think @kpierre13 @cjllanwarne?